net-mgmt/pfSense-pkg-arpwatch/files/usr/local/pkg/arpwatch.inc

```diff
@@ -0,0 +1,239 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+define('ARPWATCH_LOCAL_DIR', '/usr/local/arpwatch');
+define('ARPWATCH_ETHERCODES_URL', 'http://linuxnet.ca/ieee/oui/ethercodes.dat');
+define('ARPWATCH_SENDMAIL_PATH', '/usr/sbin/sendmail');
+define('ARPWATCH_SENDMAIL_PROXY', '/usr/local/arpwatch/sendmail_proxy.php');
+
+function arpwatch_install_command() {
+	global $config;
+
+	conf_mount_rw();
+
+	// sendmail proxy
+	arpwatch_install_sendmail_proxy();
+
+	$arpwatch_config = $config['installedpackages']['arpwatch']['config'][0];
+	$active_interfaces = $arpwatch_config['active_interfaces'];
+	$notifications_recipient = $arpwatch_config['notifications_recipient'];
+	$disable_bogons = ($arpwatch_config['disable_bogons'] == 'on') ? '-N' : '';
+	$disable_zero = ($arpwatch_config['disable_zero'] == 'on') ? '-z' : '';
+	$update_vendors = ($arpwatch_config['update_vendors'] == 'on');
+	$clear_database = ($arpwatch_config['clear_database'] == 'on');
+
+	if ($update_vendors) {
+		arpwatch_update_vendors();
+	}
+
+	if ($clear_database) {
+		arpwatch_clear_database();
+	}
+
+	$rc = [];
+	$rc['file'] = 'arpwatch.sh';
+
+	$rc['start'] = '';
+
+	foreach (explode(',', $active_interfaces) as $active_interface) {
+		$ifname = convert_friendly_interface_to_real_interface_name($active_interface);
+		$arp_file = arpwatch_get_arp_file($ifname);
+
+		touch($arp_file);
+
+		$rc['start'] .= "/usr/local/sbin/arpwatch $disable_bogons $disable_zero -f $arp_file -i $ifname -m $notifications_recipient"."\n";
+	}
+
+	$rc['stop'] = '/usr/bin/killall arpwatch || exit 0';
+
+	write_rcfile($rc);
+
+	conf_mount_ro();
+	
+	if ($arpwatch_config['enable']) {
+		restart_service("arpwatch");
+	}
+	else {
+		stop_service("arpwatch");
+	}
+}
+
+function arpwatch_deinstall_command() {
+	arpwatch_clear_database();
+	arpwatch_uninstall_sendmail_proxy();
+}
+
+function arpwatch_validate_input($post, &$input_errors) {
+	if (!isset($post['enable']) || $post['enable'] != 'on') {
+		return;
+	}
+
+	foreach ($post['active_interfaces'] as $active_interface) {
+		$realif = get_real_interface($active_interface);
+		$ip = find_interface_ip($realif);
+
+		if (!is_ipaddrv4($ip)) {
+			$ifdescr = convert_friendly_interface_to_friendly_descr($active_interface);
+
+			$input_errors[] = gettext("{$ifdescr} interface has no IPv4 address assigned.");
+		}
+	}
+
+	if (filter_var($post['notifications_recipient'], FILTER_VALIDATE_EMAIL) === false) {
+		$input_errors[] = gettext("{$post['notifications_recipient']} is not a valid email address.");
+	}
+}
+
+function arpwatch_resync_config_command() {
+	arpwatch_install_command();
+}
+
+function arpwatch_get_arp_file($ifname) {
+	return ARPWATCH_LOCAL_DIR."/arp_$ifname.dat";
+}
+
+function arpwatch_update_vendors() {
+	mwexec("/usr/bin/fetch -o ".ARPWATCH_LOCAL_DIR."/ethercodes.dat ".ARPWATCH_ETHERCODES_URL);
```

Comment:
There is a function called `download_file()` in pfsense-utils.inc. It's better to use it